lib/nxt_pipeline/pipeline.rb

```diff
@@ -44,61 +51,128 @@ def raise_duplicate_default_constructor
+        # p.step Service, constructor: ->(step, **changes) { ... }
+        if constructor.respond_to?(:call)
+          resolved_constructor = constructor
+        else
+          # p.step Service, constructor: :service
+          resolved_constructor = constructors.fetch(constructor) {
+            ::NxtPipeline.constructor(constructor) || (raise ArgumentError, "No constructor defined for #{constructor}")
+          }
+        end
+      elsif block_given?
+        # p.step :inline do ... end
+        resolved_constructor = block
       else
-        constructor = step_resolvers.lazy.map do |resolver|
-          resolver.call(argument)
+        # If no constructor was given try to resolve one
+        resolvers = constructor_resolvers.any? ? constructor_resolvers : []
+
+        constructor_from_resolvers = resolvers.map do |resolver|
+          resolver.call(argument, **opts)
         end.find(&:itself)
 
-        if constructor
-          constructor && constructors.fetch(constructor) { raise KeyError, "No step :#{argument} registered" }
-        elsif default_constructor
-          argument ||= default_constructor_name
-          default_constructor
+        # resolved constructor is a proc
+        if constructor_from_resolvers.is_a?(Proc)
+          resolved_constructor = constructor_from_resolvers
+        elsif constructor_from_resolvers.present?
+          resolved_constructor = constructors[constructor_from_resolvers]
         else
-          raise StandardError, "Could not resolve step from: #{argument}"
+          # try to resolve constructor by argument --> #TODO: Is this a good idea?
+          resolved_constructor = constructors[argument]
+        end
+
+
+        # if still no constructor resolved
+        unless resolved_constructor.present?
+          if argument.is_a?(NxtPipeline::Pipeline)
+            pipeline_constructor = ->(changes) { argument.call(changes) }
+            resolved_constructor = pipeline_constructor
+          # last chance: default constructor
+          elsif default_constructor.present?
+            resolved_constructor = default_constructor
+          # now we really give up :-(
+          else
+            raise ArgumentError, "Could neither find nor resolve any constructor for #{argument}, #{opts}"
+          end
         end
       end
 
-      register_step(argument, constructor, callbacks, **opts)
+      register_step(argument, resolved_constructor, callbacks, **opts)
     end
 
-    def execute(**change_set, &block)
+    def call(acc, &block)
       reset
 
       configure(&block) if block_given?
-      callbacks.run(:before, :execution, change_set)
+      callbacks.run(:before, :execution, acc)
 
-      result = callbacks.around :execution, change_set do
-        steps.inject(change_set) do |set, step|
-          execute_step(step, **set)
+      self.result = callbacks.around :execution, acc do
+        steps.inject(acc) do |changes, step|
+          self.result = call_step(step, changes)
```

Comment:
Why do we assign `result` after each step and after all steps?

Reply:
I think the idea was to basically have the access to the result based on the current state.